Include ignored source as part of loading field values in ValueSourceReaderOperator via BlockSourceReader.

martijnvg opened the pull request:
Currently, in compute engine when loading source if source mode is synthetic, the synthetic source loader is already used. But the ignored_source field isn't always marked as a required source field, causing the source to potentially miss a lot of fields.

This change includes `_ignored_source` field as a required stored field and allowing keyword fields without doc values or stored fields to be used in case of synthetic source.

Relying on synthetic source to get the values (because a field doesn't have stored fields / doc values) is slow. In case of synthetic source we already keep ignored field/values in a special place, named ignored source. Long term in case of synthetic source we should only load ignored source in case a field has no doc values or stored field. Like is being explored in #114886 Thereby avoiding synthesizing the complete _source in order to get only one field.

martijnvg committed:
Support reading ignored source as part of value source loading via block loaders.

Currently, in compute engine when loading source if source mode is synthetic, the synthetic source loader is already used. But the ignored_source field isn't always marked as a required source field, causing the source to potentially miss a lot of fields.

This change includes `_ignored_source` field as a required stored field.

Long term in case of synthetic source we should only load ignored source in case a field has no doc values or stored field. Like is being explored in #114886

martijnvg committed:
iter

martijnvg committed:
alter unit test

martijnvg committed:
spotless

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -22,13 +22,23 @@
  */
 public abstract class BlockSourceReader implements BlockLoader.RowStrideReader {
+
+    // _ignored_source is needed ofr synthetic source is needed for, in case stored source (default) is used,
```
nit: typo

dnhatn commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -22,13 +22,23 @@
+
+    // _ignored_source is needed ofr synthetic source is needed for, in case stored source (default) is used,
+    // then it just doesn't get loaded.
+    static final StoredFieldsSpec NEEDS_SOURCE_AND_IGNORED_SOURCE = new StoredFieldsSpec(
```
Ideally, we should avoid requesting _source and only read `_ignored_source` when synthetic source is enabled. However, we should get this in to ensure correctness, and make the blockloader more selective in a follow-up. I can help with that.

martijnvg replied:
A few unit compute engine tests failed, and instead of adjusting the tests, I made the block loaders a little more selective, based on whether synthetic source is used. See: 6702a2e7634240a5db878bf00284106d53f72288

dnhatn reviewed (approved):
LGTM. Thanks @martijnvg

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_synthetic_source_load_ignored_source

martijnvg committed:
Add sourceMode to SourceBlockLoader so that the right StoredFieldsSpec can be selected.

martijnvg committed:
fixed comment

martijnvg committed:
fixed field mapper tests

martijnvg committed:
alter test based on previous commit

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_synthetic_source_load_ignored_source

martijnvg committed:
added a random test that groups by a field that could be disabled

dnhatn reviewed (approved):
The new changes look good. The failing test appears to be relevant.

````
StandardVersusLogsIndexModeChallengeRestIT > testEsqlTermsAggregationByMethod FAILED
    org.elasticsearch.client.WarningFailureException: method [POST], host [http://[::1]:38381], URI [/_query], status line [HTTP/1.1 200 OK]
    Warnings: [Field [method] cannot be retrieved, it is unsupported or not indexed; returning null]
    {"took":39,"columns":[{"name":"count(*)","type":"long"},{"name":"method","type":"text"}],"values":[[129,null]]}
        at __randomizedtesting.SeedInfo.seed([7C34733CCD9E9B9A:AB344202DE16E96D]:0)
        at app//org.elasticsearch.client.RestClient.convertResponse(RestClient.java:347)
        at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:317)
        at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)
        at app//org.elasticsearch.datastreams.logsdb.qa.AbstractChallengeRestTest.esql(AbstractChallengeRestTest.java:290)
        at app//org.elasticsearch.datastreams.logsdb.qa.AbstractChallengeRestTest.esqlContender(AbstractChallengeRestTest.java:284)
        at app//org.elasticsearch.datastreams.logsdb.qa.StandardVersusLogsIndexModeChallengeRestIT.testEsqlTermsAggregationByMethod(StandardVersusLogsIndexModeChallengeRestIT.java:317)
```

martijnvg committed:
Take into account text fields with keyword sub fields.
If text fields are not stored, then keyword sub fields
can be used to syntesize fields for text parent field.

martijnvg commented:
> The failing test appears to be relevant.

Yes, the randomized test failure showed that `TextFieldType#blockLoader(...)` wasn't taking into account keyword sub fields if a text field didn't store source. I pushed c6e303effd154549f0ebdfe7c1e821ad1146d9fe to address this.

martijnvg committed:
Merge remote-tracking branch 'es/main' into esql_synthetic_source_load_ignored_source

martijnvg commented:
Looks like the CI check here is stuck. All PR CI did complete successfully.

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/115064)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/115064">](https://github.com/elastic/elasticsearch/pull/115064) |
